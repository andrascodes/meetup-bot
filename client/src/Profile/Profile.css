.ProfileForm {
  width: 50%;
}

@media only screen and (max-width : 768px) {
  .ProfileForm {
    width: 100%;
  }
}