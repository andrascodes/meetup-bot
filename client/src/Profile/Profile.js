import React, { Component } from 'react'
import { Redirect } from 'react-router-dom'

import { Error } from '../Error'
import { createAuthApi } from '../lib/authApi'

import './Profile.css'

export class Profile extends Component {

  state = {
    username: (localStorage.username || ''),
    email: (localStorage.email && (localStorage.email === 'null' ? '' : localStorage.email)) || '',
    password: '',
    error: false
  }

  putUserProfile = createAuthApi(fetch).putUserProfile

  handleUsernameChange = e => {
    e.persist()
    e.preventDefault()
    this.setState(state => ({
      username: e.target.value
    }))
  }

  handleEmailChange = e => {
    e.persist()
    e.preventDefault()
    this.setState(state => ({
      email: e.target.value
    }))
  }

  handlePasswordChange = e => {
    e.persist()
    e.preventDefault()
    this.setState(state => ({
      password: e.target.value
    }))
  }

  handleSubmit = e => {
    e.persist()
    e.preventDefault()

    this.putUserProfile(this.state.username, this.state.password, this.state.email)
    .then(res => this.setState(state => ({
      error: {
        status: 200,
        statusText: 'changed'
      }
    })))
    .catch(error => {
      console.log(error)
      this.setState(state => ({ error }))
    })
  }

  render() {
    if(this.state.error.status === 401 || this.state.error.statusText === 'changed') {
      // Unauthorized
      localStorage.clear()
      return (
        <Redirect to={{
          pathname: '/login',
          state: {
            from: this.props.location
          }
        }}/>
      )
    }
    else if(this.state.error.status === 400) {
      console.log(this.state.error)
      return (
        <div className="Conversations">
          <Error text={`Email validation failed.`}/>
        </div>
      )
    }
    else if(this.state.error) {
      return (
        <div className="Conversations">
          <Error />
        </div>
      )
    }
    else {
      return (
        <div className="Profile">
          <div className="container-fluid">
            <h1>Profile</h1>
            <form className="ProfileForm" role="form" onSubmit={this.handleSubmit}>
              <fieldset>
                <div className="form-group">
                  <label htmlFor="emailInput">Email</label>
                  <input 
                    type="email" 
                    className="form-control" 
                    id="emailInput"
                    placeholder="Email" 
                    defaultValue={this.state.email}
                    onChange={this.handleEmailChange}
                  />
                </div>

                <div className="form-group">
                  <label htmlFor="usernameInput">Username</label>
                  <input 
                    type="text" 
                    className="form-control" 
                    id="usernameInput" 
                    placeholder="Username"
                    defaultValue={this.state.username}
                    onChange={this.handleUsernameChange}
                  />
                </div>
                
                <div className="form-group">
                  <label htmlFor="passwordInput">New Password</label>
                  <input 
                    type="password" 
                    className="form-control" 
                    id="passwordInput" 
                    placeholder="Password" 
                    onChange={this.handlePasswordChange}
                  />
                </div>
                
                <button type="submit" className="btn btn-default">Submit</button>
                <div className="ProfileTip">
                  <small>You will have to sign in again, after submitting this form.</small>
                </div>
              </fieldset>
            </form>
          </div>
        </div>
      )
    }
  }
}