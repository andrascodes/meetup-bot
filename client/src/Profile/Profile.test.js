import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import { Profile } from './Profile'

describe(`Profile`, () => {

  beforeEach(() => {
    global.localStorage = {
      clear: jest.fn(),
      username: 'admin',
      email: 'admin@mail.com'
    }
  })

  test(`renders Redirect to /login when state.error.status is 401: Unauthorized`, () => {
    const wrapper = mount(
      <MemoryRouter>
        <Profile error={{ status: 401 }}/>
      </MemoryRouter>
    )
    expect(global.localStorage.clear).toHaveBeenCalled()
    expect(wrapper.find('Redirect').prop('to').pathname).toEqual('/login')
  })

  test(`renders Redirect to /login when state.statusText is 'changed'`, () => {
    const wrapper = mount(
      <MemoryRouter>
        <Profile error={{ status: 200, statusText: 'changed' }}/>
      </MemoryRouter>
    )
    expect(global.localStorage.clear).toHaveBeenCalled()
    expect(wrapper.find('Redirect').prop('to').pathname).toEqual('/login')
  })
  
  test(`renders Error when state.error is defined to anything else`, () => {
    const wrapper = mount(
      <Profile error={{ status: 500 }}/>
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('renders correctly', () => {
    const wrapper = mount(<Profile />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('renders the username when state.username is changed', () => {
    const wrapper = mount(<Profile />)
   
    const username = 'user_1'
    wrapper.find('#usernameInput').simulate('change', { 
      persist: () => {}, 
      preventDefault: () => {},
      target: {
        value: username,
      }
    })
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  
  test('renders the email when state.email is changed', () => {
    const wrapper = mount(<Profile />)
   
    const email = 'email@mail.com'
    wrapper.find('#emailInput').simulate('change', { 
      persist: () => {}, 
      preventDefault: () => {},
      target: {
        value: email,
      }
    })
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('renders when state.password is changed', () => {
    const wrapper = mount(<Profile />)
   
    const password = 'another_password'
    wrapper.find('#passwordInput').simulate('change', { 
      persist: () => {}, 
      preventDefault: () => {},
      target: {
        value: password,
      }
    })
    expect(toJson(wrapper)).toMatchSnapshot()
  })

})