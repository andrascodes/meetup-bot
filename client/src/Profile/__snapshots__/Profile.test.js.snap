exports[`Profile renders Error when state.error is defined to anything else 1`] = `
<Profile
  error={
    Object {
      "status": 500,
    }
  }>
  <div
    className="Conversations">
    <Component>
      <div
        className="container-fluid">
        <h2>
          Error. 
          Try again later!
        </h2>
      </div>
    </Component>
  </div>
</Profile>
`;

exports[`Profile renders correctly 1`] = `
<Profile>
  <div
    className="Profile">
    <div
      className="container-fluid">
      <h1>
        Profile
      </h1>
      <form
        className="ProfileForm"
        onSubmit={[Function]}
        role="form">
        <fieldset>
          <div
            className="form-group">
            <label
              htmlFor="emailInput">
              Email
            </label>
            <input
              className="form-control"
              defaultValue="admin@mail.com"
              id="emailInput"
              onChange={[Function]}
              placeholder="Email"
              type="email" />
          </div>
          <div
            className="form-group">
            <label
              htmlFor="usernameInput">
              Username
            </label>
            <input
              className="form-control"
              defaultValue="admin"
              id="usernameInput"
              onChange={[Function]}
              placeholder="Username"
              type="text" />
          </div>
          <div
            className="form-group">
            <label
              htmlFor="passwordInput">
              New Password
            </label>
            <input
              className="form-control"
              id="passwordInput"
              onChange={[Function]}
              placeholder="Password"
              type="password" />
          </div>
          <button
            className="btn btn-default"
            type="submit">
            Submit
          </button>
          <div
            className="ProfileTip">
            <small>
              You will have to sign in again, after submitting this form.
            </small>
          </div>
        </fieldset>
      </form>
    </div>
  </div>
</Profile>
`;

exports[`Profile renders the email when state.email is changed 1`] = `
<Profile>
  <div
    className="Profile">
    <div
      className="container-fluid">
      <h1>
        Profile
      </h1>
      <form
        className="ProfileForm"
        onSubmit={[Function]}
        role="form">
        <fieldset>
          <div
            className="form-group">
            <label
              htmlFor="emailInput">
              Email
            </label>
            <input
              className="form-control"
              defaultValue="email@mail.com"
              id="emailInput"
              onChange={[Function]}
              placeholder="Email"
              type="email" />
          </div>
          <div
            className="form-group">
            <label
              htmlFor="usernameInput">
              Username
            </label>
            <input
              className="form-control"
              defaultValue="admin"
              id="usernameInput"
              onChange={[Function]}
              placeholder="Username"
              type="text" />
          </div>
          <div
            className="form-group">
            <label
              htmlFor="passwordInput">
              New Password
            </label>
            <input
              className="form-control"
              id="passwordInput"
              onChange={[Function]}
              placeholder="Password"
              type="password" />
          </div>
          <button
            className="btn btn-default"
            type="submit">
            Submit
          </button>
          <div
            className="ProfileTip">
            <small>
              You will have to sign in again, after submitting this form.
            </small>
          </div>
        </fieldset>
      </form>
    </div>
  </div>
</Profile>
`;

exports[`Profile renders the username when state.username is changed 1`] = `
<Profile>
  <div
    className="Profile">
    <div
      className="container-fluid">
      <h1>
        Profile
      </h1>
      <form
        className="ProfileForm"
        onSubmit={[Function]}
        role="form">
        <fieldset>
          <div
            className="form-group">
            <label
              htmlFor="emailInput">
              Email
            </label>
            <input
              className="form-control"
              defaultValue="admin@mail.com"
              id="emailInput"
              onChange={[Function]}
              placeholder="Email"
              type="email" />
          </div>
          <div
            className="form-group">
            <label
              htmlFor="usernameInput">
              Username
            </label>
            <input
              className="form-control"
              defaultValue="user_1"
              id="usernameInput"
              onChange={[Function]}
              placeholder="Username"
              type="text" />
          </div>
          <div
            className="form-group">
            <label
              htmlFor="passwordInput">
              New Password
            </label>
            <input
              className="form-control"
              id="passwordInput"
              onChange={[Function]}
              placeholder="Password"
              type="password" />
          </div>
          <button
            className="btn btn-default"
            type="submit">
            Submit
          </button>
          <div
            className="ProfileTip">
            <small>
              You will have to sign in again, after submitting this form.
            </small>
          </div>
        </fieldset>
      </form>
    </div>
  </div>
</Profile>
`;

exports[`Profile renders when state.password is changed 1`] = `
<Profile>
  <div
    className="Profile">
    <div
      className="container-fluid">
      <h1>
        Profile
      </h1>
      <form
        className="ProfileForm"
        onSubmit={[Function]}
        role="form">
        <fieldset>
          <div
            className="form-group">
            <label
              htmlFor="emailInput">
              Email
            </label>
            <input
              className="form-control"
              defaultValue="admin@mail.com"
              id="emailInput"
              onChange={[Function]}
              placeholder="Email"
              type="email" />
          </div>
          <div
            className="form-group">
            <label
              htmlFor="usernameInput">
              Username
            </label>
            <input
              className="form-control"
              defaultValue="admin"
              id="usernameInput"
              onChange={[Function]}
              placeholder="Username"
              type="text" />
          </div>
          <div
            className="form-group">
            <label
              htmlFor="passwordInput">
              New Password
            </label>
            <input
              className="form-control"
              id="passwordInput"
              onChange={[Function]}
              placeholder="Password"
              type="password" />
          </div>
          <button
            className="btn btn-default"
            type="submit">
            Submit
          </button>
          <div
            className="ProfileTip">
            <small>
              You will have to sign in again, after submitting this form.
            </small>
          </div>
        </fieldset>
      </form>
    </div>
  </div>
</Profile>
`;
